File: java/easy/tic-tac-toe/TicTacToe.java
package tictactoe;
import java.util.Scanner;
import java.util.ArrayList;
import java.util.Arrays;


class TicTacToe {

    private char[][] matrix;

    public TicTacToe(String initial) {
        if (initial == null)
            initial = "_________";
        this.matrix = new char[3][3];
        for (int i = 0; i < 3; i++) {
            for (int j = 0; j < 3; j++) {
                matrix[i][j] = initial.charAt(i * 3 + j);
            }
        }
    }

    private boolean getWin(char symbol) {
        char[] checked = { symbol, symbol, symbol };
        ArrayList<Character> checkedList = new ArrayList<>();
        for (char ch : checked) {
            checkedList.add(ch);
        }
        ArrayList<Character> diagonal = new ArrayList<>();
        ArrayList<Character> horizontal;
        for (int i = 0; i < this.matrix.length; i++) {
            if (Arrays.equals(matrix[0], checked))
                return true;
            horizontal = new ArrayList<>();
            for (int j = 0; j < this.matrix[i].length; j++) {
                horizontal.add(this.matrix[j][i]);
                if (j + i == 2)
                    diagonal.add(matrix[i][j]);
            }
            if (horizontal.equals(checkedList))
                return true;
        }
        if (diagonal.equals(checkedList))
            return true;
        diagonal = new ArrayList<>();
        for (int i = 0; i < 3; i++) {
            diagonal.add(this.matrix[i][i]);
        }
        return diagonal.equals(checkedList);
    }

    private boolean getImpossible() {
        if (this.getWin('X') && this.getWin('O'))
            return true;
        int x = 0;
        int o = 0;
        for (int i = 0; i < this.matrix.length; i++) {
            for (int j = 0; j < this.matrix[i].length; j++) {
                if (matrix[i][j] == 'X') x++;
                if (matrix[i][j] == 'O') o++;
            }
        }
        return !((0 <= Math.abs(x - o)) && (Math.abs(x - o) <= 1));
    }

    private boolean getDraw() {
        for (int i = 0; i < this.matrix.length; i++) {
            for (int j = 0; j < this.matrix[i].length; j++) {
                if (matrix[i][j] == '_')
                    return false;
            }
        }
        return true;
    }

    private boolean check() {
        if (this.getImpossible()) {
            System.out.println("Impossible");
            return true;
        }
        else if (this.getWin('X')) {
            System.out.println("X wins");
            return true;
        }
        else if (this.getWin('O')) {
            System.out.println("O wins");
            return true;
        }
        else if (this.getDraw()) {
            System.out.println("Draw");
            return true;
        }
        else {
            return false;
        }
    }

    private boolean isOccupied(int x, int y) {
        return this.matrix[x][y] != '_';
    }

    public TicTacToe setCoordinates(char symbol) {
        while (true) {
            int x;
            int y;
            System.out.print("Enter the coordinates: ");
            Scanner scanner = new Scanner(System.in);
            try {
                y = scanner.nextInt();
                x = scanner.nextInt();
            }
            catch (Exception exc) {
                System.out.println("You should enter numbers!");
                continue;
            }
            if (1 <= x && x <= 3 && 1 <= y && y <= 3) {
                x = -x;
                --y;
            }
            else {
                System.out.println("Coordinates should be from 1 to 3!");
                continue;
            }
            if (x < 0) {
                x = x + 3;
            }
            if (y < 0) {
                y = y + 3;
            }
            if (this.isOccupied(x, y)) {
                System.out.println("This cell is occupied! Choose another one!");
                continue;
            }
            this.matrix[x][y] = symbol;
            return this;
        }
    }

    private boolean move(char symbol) {
        System.out.println(this);
        if (!this.check()) {
            this.setCoordinates(symbol);
            return false;
        }
        return true;
    }

    @Override
    public String toString() {
        StringBuilder result = new StringBuilder();
        result.append("---------\n");
        for (int i = 0; i < 3; i++) {
            result.append("|");
            for (int j = 0; j < 3; j++) {
                result.append(" ").append(matrix[i][j]);
            }
            result.append(" |\n");
        }
        result.append("---------\n");
        return result.toString();
    }

    public void main() {
        boolean haveWon = false;
        char[] symbols = { 'X', 'O' };
        while (!haveWon) {
            for (char symbol: symbols) {
                if (this.move(symbol)) {
                    haveWon = true;
                    break;
                }
            }
        }
    }
}

public class Main {
    public static void main(String[] args) {
        TicTacToe game = new TicTacToe(null);
        game.main();
    }
}

File: python/easy/tic-tac-toe/tictactoe.py
class TicTacToe:
    def __init__(self, init_string="_________"):
        self.matrix = []
        i = 0
        while i < 9:
            row = []
            for symbol in init_string[i:i + 3]:
                row.append(symbol)
            self.matrix.append(row)
            i += 3

    def __str__(self):
        result = "---------\n"
        for row in self.matrix:
            result += "| "
            result += " ".join(row)
            result += " |\n"
        result += "---------"
        return result

    def __get_win(self, symbol="X"):
        checked = [symbol] * 3
        diagonal = []
        for i, row in enumerate(self.matrix):
            if row == checked:
                return True
            horizontal = []
            for j, column in enumerate(row):
                horizontal.append(self.matrix[j][i])
                if j + i == 2:
                    diagonal.append(column)
            if horizontal == checked:
                return True
        if diagonal == checked:
            return True
        diagonal = []
        for i in range(3):
            diagonal.append(self.matrix[i][i])
        if diagonal == checked:
            return True

    def __get_impossible(self):
        if self.__get_win() and self.__get_win("O"): return True
        x = 0
        o = 0
        for row in self.matrix:
            x += row.count("X")
            o += row.count("O")
        return not (0 <= abs(x - o) <= 1)

    def __get_draw(self):
        for row in self.matrix:
            for column in row:
                if column == "_":
                    return False
        return True

    def __check(self):
        if self.__get_impossible():
            print("Impossible")
            return True
        elif self.__get_win("X"):
            print("X wins")
            return True
        elif self.__get_win("O"):
            print("O wins")
            return True
        elif self.__get_draw():
            print("Draw")
            return True
        else:
            return False

    def __is_occupied(self, x, y):
        return self.matrix[x][y] != "_"

    def __set_coordinates(self, symbol="X"):
        while True:
            try:
                y, x = input("Enter the coordinates: ").split()
            except:
                continue
            if x.isdigit() and y.isdigit():
                x = int(x)
                y = int(y)
            else:
                print("You should enter numbers!")
                continue
            if 1 <= x <= 3 and 1 <= y <= 3:
                x = -x
                y -= 1
            else:
                print("Coordinates should be from 1 to 3!")
                continue
            if self.__is_occupied(x, y):
                print("This cell is occupied! Choose another one!")
                continue
            self.matrix[x][y] = symbol
            return self

    def __move(self, symbol="X"):
        print(self)
        if not self.__check():
            self.__set_coordinates(symbol)
            return False
        return True

    def main(self):
        have_won = False
        while not have_won:
            for c in ("X", "O"):
                if self.__move(c):
                    have_won = True
                    break


if __name__ == '__main__':
    game = TicTacToe()
    game.main()
